datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java

```diff
@@ -124,25 +120,27 @@ public AbstractKafkaConnector(String connectorName, Properties config, GroupIdCo
   public synchronized void onAssignmentChange(List<DatastreamTask> tasks) {
     _logger.info("onAssignmentChange called with tasks {}", tasks);
 
-    HashSet<DatastreamTask> toCancel = new HashSet<>(_runningTasks.keySet());
-    toCancel.removeAll(tasks);
-
-    for (DatastreamTask task : toCancel) {
-      AbstractKafkaBasedConnectorTask connectorTask = _runningTasks.remove(task);
-      connectorTask.stop();
-      _taskThreads.remove(task);
-    }
+    synchronized (_runningTasks) {
+      HashSet<DatastreamTask> toCancel = new HashSet<>(_runningTasks.keySet());
+      toCancel.removeAll(tasks);
 
-    for (DatastreamTask task : tasks) {
-      AbstractKafkaBasedConnectorTask kafkaBasedConnectorTask = _runningTasks.get(task);
-      if (kafkaBasedConnectorTask != null) {
-        kafkaBasedConnectorTask.checkForUpdateTask(task);
-        // make sure to replace the DatastreamTask with most up to date info
-        _runningTasks.put(task, kafkaBasedConnectorTask);
-        continue; // already running
+      for (DatastreamTask task : toCancel) {
+        ConnectorTaskEntry connectorTaskEntry = _runningTasks.remove(task);
+        connectorTaskEntry.getConnectorTask().stop();
       }
 
-      createKafkaConnectorTask(task);
+      for (DatastreamTask task : tasks) {
+        ConnectorTaskEntry connectorTaskEntry = _runningTasks.get(task);
+        if (connectorTaskEntry != null) {
+          AbstractKafkaBasedConnectorTask kafkaBasedConnectorTask = connectorTaskEntry.getConnectorTask();
+          kafkaBasedConnectorTask.checkForUpdateTask(task);
+          // make sure to replace the DatastreamTask with most up to date info
+          _runningTasks.put(task, new ConnectorTaskEntry(kafkaBasedConnectorTask, connectorTaskEntry.getThread()));
```

Comment:
let's discuss this in person, I'm a little confused.

Reply:
For (3): Dependencies are not a worry at the moment. Only partition-managed datastreams use dependencies, and when a new task is created which has a list of dependencies, the taskPrefix of the new task is always different. This may become a concern in the future we change the semantics of this, where tasks having the same task prefix can have different dependencies.